File: backend/controllers/bookingController.js
// cortate/backend/controllers/bookingController.js

const Booking = require('../models/Booking');
const Barber = require('../models/Barber');
const Penalty = require('../models/Penalty');

// Crear nueva reserva
const createBooking = async (req, res) => {
	try {
		const {
			barbero,
			servicio,
			monto,
			fechaReserva,
			horaReserva,
			tipoAtencion
		} = req.body;

		const nuevaReserva = new Booking({
			cliente: req.user.id,
			barbero,
			servicio,
			monto,
			fechaReserva,
			horaReserva,
			tipoAtencion
		});

		await nuevaReserva.save();

		res.status(201).json({ msg: 'Reserva creada. A la espera de confirmación.', reserva: nuevaReserva });
	} catch (err) {
		console.error(err);
		res.status(500).json({ msg: 'Error al crear la reserva.' });
	}
};

// Confirmar manualmente (barbero)
const confirmarReserva = async (req, res) => {
	try {
		const reserva = await Booking.findById(req.params.id);
		if (!reserva) return res.status(404).json({ msg: 'Reserva no encontrada.' });

		if (reserva.barbero.toString() !== req.user.id)
			return res.status(401).json({ msg: 'No autorizado.' });

		reserva.estado = 'confirmado';
		reserva.confirmadoManual = true;

		await reserva.save();

		res.status(200).json({ msg: 'Reserva confirmada.' });
	} catch (err) {
		console.error(err);
		res.status(500).json({ msg: 'Error al confirmar reserva.' });
	}
};

// Cancelar reserva
const cancelarReserva = async (req, res) => {
	try {
		const reserva = await Booking.findById(req.params.id);
		if (!reserva) return res.status(404).json({ msg: 'Reserva no encontrada.' });

		const esCliente = reserva.cliente.toString() === req.user.id;
		const esBarbero = reserva.barbero.toString() === req.user.id;

		if (!esCliente && !esBarbero)
			return res.status(401).json({ msg: 'No autorizado.' });

		reserva.estado = 'cancelado';
		reserva.canceladoPor = esCliente ? 'cliente' : 'barbero';
		await reserva.save();

		// Penalización si se cancela tarde o se rechaza
		if (reserva.estado === 'pendiente' || reserva.confirmadoManual) {
			await new Penalty({
				usuario: req.user.id,
				tipoUsuario: req.user.role,
				razon: 'cancelacionTardia',
				booking: reserva._id
			}).save();
		}

		res.status(200).json({ msg: 'Reserva cancelada.' });
	} catch (err) {
		console.error(err);
		res.status(500).json({ msg: 'Error al cancelar reserva.' });
	}
};

// Ver reservas de usuario (cliente)
const getUserBookings = async (req, res) => {
	try {
		const reservas = await Booking.find({ cliente: req.user.id })
			.populate('barbero')
			.sort({ fechaReserva: -1 });

		res.status(200).json(reservas);
	} catch (err) {
		console.error(err);
		res.status(500).json({ msg: 'Error al obtener reservas.' });
	}
};

// Ver reservas para barbero
const getBarberBookings = async (req, res) => {
	try {
		const reservas = await Booking.find({ barbero: req.user.id })
			.populate('cliente')
			.sort({ fechaReserva: -1 });

		res.status(200).json(reservas);
	} catch (err) {
		console.error(err);
		res.status(500).json({ msg: 'Error al obtener reservas del barbero.' });
	}
};

module.exports = {
	createBooking,
	confirmarReserva,
	cancelarReserva,
	getUserBookings,
	getBarberBookings
};

File: backend/controllers/bookingController.py
# cortate/backend/controllers/bookingController.py

from flask import Blueprint, request, jsonify
from bson import ObjectId
from config.database import bookings_collection

booking_controller = Blueprint('booking_controller', __name__)

# Crear reserva
@booking_controller.route('/create', methods=['POST'])
def crear_reserva():
    data = request.get_json()
    if not data:
        return jsonify({"error": "Datos incompletos"}), 400
    resultado = bookings_collection.insert_one(data)
    data["_id"] = str(resultado.inserted_id)
    return jsonify(data), 201

# Obtener todas las reservas
@booking_controller.route('/all', methods=['GET'])
def listar_reservas():
    reservas = []
    for r in bookings_collection.find():
        r["_id"] = str(r["_id"])
        reservas.append(r)
    return jsonify(reservas), 200

# Obtener reservas de un cliente
@booking_controller.route('/cliente/<cliente_id>', methods=['GET'])
def reservas_cliente(cliente_id):
    reservas = []
    for r in bookings_collection.find({"cliente_id": cliente_id}):
        r["_id"] = str(r["_id"])
        reservas.append(r)
    return jsonify(reservas), 200

# Obtener reservas de un barbero
@booking_controller.route('/barbero/<barbero_id>', methods=['GET'])
def reservas_barbero(barbero_id):
    reservas = []
    for r in bookings_collection.find({"barbero_id": barbero_id}):
        r["_id"] = str(r["_id"])
        reservas.append(r)
    return jsonify(reservas), 200

# Cancelar reserva
@booking_controller.route('/delete/<reserva_id>', methods=['DELETE'])
def cancelar_reserva(reserva_id):
    resultado = bookings_collection.delete_one({"_id": ObjectId(reserva_id)})
    if resultado.deleted_count == 0:
        return jsonify({"error": "Reserva no encontrada"}), 404
    return '', 204
